fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/SumoAmbassador.java

```diff
@@ -249,25 +244,54 @@ private void receiveInteraction(VehicleRoutesInitialization interaction) throws
     private synchronized void receiveInteraction(VehicleTypesInitialization interaction) throws InternalFederateException {
         log.debug("Received VehicleTypesInitialization");
 
-        if (cachedVehicleRoutesInitialization != null && descriptor != null) {
-            writeRouteFile(interaction, cachedVehicleRoutesInitialization);
-            startSumoLocal();
-            initTraci();
-            completeRoutes();
-        }
         cachedVehicleTypesInitialization = interaction;
+        if (sumoReadyToStart()) {
+            sumoStartupProcedure();
+        }
     }
 
-    private void completeRoutes() throws InternalFederateException {
+    private boolean sumoReadyToStart() {
+        return descriptor != null && cachedVehicleRoutesInitialization != null && cachedVehicleTypesInitialization != null;
+    }
+
+    protected void sumoStartupProcedure() throws InternalFederateException {
+        writeTypesFromMapping(cachedVehicleTypesInitialization);
+        startSumoLocal();
+        initTraci();
+        readInitialRoutesFromTraci();
+        addInitialRoutesFromMapping();
+    }
+
+    /**
+     * Read Routes priorly defined in Sumo route-file to later make them available to the rest of the
+     * simulations using {@link VehicleRouteRegistration}.
+     *
+     * @throws InternalFederateException if Traci connection couldn't be established
+     */
+    private void readInitialRoutesFromTraci() throws InternalFederateException {
         for (String id : traci.getRouteControl().getRouteIds()) {
-            if (!sumoKnownRoutes.contains(id)) {
+            if (!routeCache.containsKey(id)) {
                 VehicleRoute route = readRouteFromTraci(id);
-                sumoKnownRoutes.add(route.getId());
                 routeCache.put(route.getId(), route);
             }
         }
     }
 
+    /**
+     * Passes on the routes from Mapping to SUMO.
+     *
+     * @throws InternalFederateException if there was a problem with traci
+     */
+    private void addInitialRoutesFromMapping() throws InternalFederateException {
```

Comment:
Routes do not come from mapping, but from ApplicationAmbassador. In fact, they could come from everything which is coupled to the RTI. Just call it `addInitialRoutes`.

Reply:
👍